core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -104,7 +105,8 @@
     private static final List<RequestConverterFunction> defaultRequestConverters =
             ImmutableList.of(new JacksonRequestConverterFunction(),
                              new StringRequestConverterFunction(),
-                             new ByteArrayRequestConverterFunction());
+                             new ByteArrayRequestConverterFunction(),
+                             new JavaBeanRequestConverterFunction());
```

Comment:
I guess it would be used more than `StringRequestConverterFunction` and `ByteArrayRequestConverterFunction`.  😃 How about moving this to up?

Reply:
Actually, I add `JavaBeanRequestConverterFunction` as the last item in this `defaultRequestConverters` list **on purpose**, to make it has the **lowest** priority in all the converters, so `ByteArray`/`String`/`Jackson` requests will be handled by corresponding converters first, and then pass all the other requests to `JavaBeanRequestConverterFunction`, which will treat request object as Java Bean.
By this way, I don't need to implement a complicated `shouldHandle this request or should pass it to other converts?` check  for `JavaBeanRequestConverterFunction`

But to be honest, this trick is a little too hacky, and hard to understand.

So, I will change it, to move `JavaBeanRequestConverterFunction` to the head of `defaultRequestConverters` list, and implement a better `shouldHandle...?` check for it.

Reply:
I see. I'm not sure but what if we instantiate a request converter which is dedicated to a class of Java bean. In [this comment](https://github.com/line/armeria/pull/1096#discussion_r175024295), I mentioned a singleton instance, but it may be better to create a new converter instance on demand.

```
public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
    
    private final Class<?> expectedResultType;

    public JavaBeanRequestConverterFunction(Class<?> expectedResultType) {
        this.expectedResultType = expectedResultType;
    }

    @Override
    public Object convertRequest(final ServiceRequestContext ctx,
                                 final AggregatedHttpMessage request,
                                 final Class<?> expectedResultType) throws Exception {
        if (this.expectedResultType != expectedResultType) {
            return RequestConverterFunction.fallthrough();
        }
        ...
    }
```

Reply:
I think I can do this:
1. create a singleton instance of `JavaBeanRequestConverterFunction`
2. call `checkResultType` method of this singleton instance, for **every** `@RequestObject` annotated method params, at the server starts up phase.
3. in this `checkResultType` method, `JavaBeanRequestConverterFunction` can check if `@Param/@Header` annotated fields exist? if so, build a `BeanParamInfo` for this ResultType, and add this  ResultType in to a `tagetResultTypeList`
4. `shouldHandle this request?` check can simply check if `expectedResultType` exists in `tagetResultTypeList`
5. if `tagetResultTypeList` is empty, which means there is no JavaBean request param exists, so we can remove `JavaBeanRequestConverterFunction` from our `defaultRequestConverters` list 

How do you think about this strategy?

Reply:
fixed, but a little different from the plan I mentioned above.
1. not using `singleton instance` pattern, but change `beanParamsMap` to `static`
2. created a `static void initFor(final Class<?> expectedResultType)`  method, to check if every result type should be processed, and if so, build a `BeanParamInfo` for it, and save to `beanParamsMap`. This method is invoked at the server starts up phase, for every `@RequestObject` annotated method params
3. `shouldHandle this request?` check simply check if `expectedResultType` is in the `beanParamsMap`
4. abandoned the plan which remove the `JavaBeanRequestConverterFunction` instance from  `defaultRequestConverters` list. because there is a Immutable copy of this list in every `AnnotatedHttpServiceMethod`, so the remove logic will be a little complicated. And, `shouldHandle this request?` is very quick, so I think there will be no performance issue even we don't remove `JavaBeanRequestConverterFunction` which is no need to exist